### PR TITLE
feat(array): refactor `StreamChunk` & introduce record iterator 

### PR DESCRIPTION
**Pull request by BugenZhao**
## What's changed and what's your intention?

In this PR, we refactor `StreamChunk` with `ops` and `data_chunk` instead of unzipping the data chunk into parts, so that we can reuse some functionalities of `DataChunk`. 

https://github.com/singularity-data/risingwave/blob/1228f9272c6e7bcce9db86f7f7c982a1bc5032e0/src/common/src/array/stream_chunk.rs#L72-L79

Then, we remove the `StreamChunkRowRef` introduced in #1813 and provide a record iterator based on `DataChunkRowRef` whose item is...

https://github.com/singularity-data/risingwave/blob/1cc6400140e40843de46f79f2a575da6989a206d/src/common/src/array/stream_chunk_iter.rs#L74-L82

Some executors can be benefited from this record-style interface. For now, we keep the row interface for compatibility.

### Benchmark
```
test array::stream_chunk_iter::tests::bench_record_iterator                 ... bench:      45,728 ns/iter (+/- 1,268)
test array::stream_chunk_iter::tests::bench_rows_iterator                   ... bench:      39,444 ns/iter (+/- 865)
test array::stream_chunk_iter::tests::bench_rows_iterator_from_records      ... bench:      73,246 ns/iter (+/- 2,349)
test array::stream_chunk_iter::tests::bench_rows_iterator_vec_of_datum_refs ... bench:     254,524 ns/iter (+/- 11,879)
```

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
- #1839
- #1840

### CODE REVIEW

File: src/common/src/array/stream_chunk_iter.rs [hunk 1]
```diff
@@ -12,137 +12,97 @@
...
+            inner: self.data_chunk.rows(),
         }
     }
+
+    /// Return an iterator on rows of this stream chunk.
+    ///
+    /// Should consider using [`StreamChunk::records`] if possible.
+    pub fn rows(&self) -> impl Iterator<Item = (Op, RowRef<'_>)> {
+        self.records().flat_map(|r| r.into_row_refs())
+    }
```
Comment (yuhao-su) [inline src/common/src/array/stream_chunk_iter.rs:36]: There are additional cost when `.record()` and `flat_map` it back. Can we simply iter the data chunk?

Comment (BugenZhao) [reply src/common/src/array/stream_chunk_iter.rs]: Fixed.

**Comment by yuhao-su** [comment]
LGTM!

**Review by yuhao-su — APPROVED** [review]

File: src/common/src/array/stream_chunk_iter.rs [hunk 2]
```diff
@@ -12,140 +12,177 @@
...
+        let op = unsafe { self.chunk.ops().get_unchecked(row.index()) };
+
+        match op {
+            Op::Insert => Some(RecordRef::Insert(row)),
+            Op::Delete => Some(RecordRef::Delete(row)),
+            Op::UpdateDelete => {
+                let insert_row = self.inner.next().expect("expect a row after U-");
+                // SAFETY: index is checked since `insert_row` is `Some`.
+                let op = unsafe { *self.chunk.ops().get_unchecked(insert_row.index()) };
+                debug_assert_eq!(op, Op::UpdateInsert, "expect a U+ after U-");
```
Comment (fuyufjh) [inline src/common/src/array/stream_chunk_iter.rs:68]:
```suggestion
                debug_assert_eq!(op, Op::UpdateInsert, "expect an U+ after U-");
```

File: src/common/src/array/stream_chunk.rs [hunk 3]
```diff
@@ -75,27 +74,18 @@ pub type Ops<'a> = &'a [Op];
 pub struct StreamChunk {
     // TODO: Optimize using bitmap
     ops: Vec<Op>,
-    columns: Vec<Column>,
-    visibility: Option<Bitmap>,
-    cardinality: usize,
+
+    pub(super) data_chunk: DataChunk,
```
Comment (fuyufjh) [inline src/common/src/array/stream_chunk.rs:78]: How about simply call it 'data'?

**Review by fuyufjh — COMMENTED** [review]
Generally LGTM. 

And a question: what's the benefit from doing so?

**Comment by BugenZhao** [comment]
> And a question: what's the benefit from doing so?

Actually the main purpose of this PR is to reuse the row-level iterator of data chunk in #1905 for stream chunk. For `record`, some executors may use this interface to handle "update" more elegantly, but in the future. 😁

File: src/common/src/array/stream_chunk_iter.rs (hunk 2 above)
Comment (BugenZhao) [reply src/common/src/array/stream_chunk_iter.rs:68]: Should be "a" here since we pronounce it as `[juː]`.

**Review by skyzh — APPROVED** [review]

**Review by fuyufjh — APPROVED** [review]

**Pull request merged**